WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/PostLikeActionUseCase.kt

```diff
@@ -0,0 +1,84 @@
+package org.wordpress.android.ui.reader.repository.usecases
+
+import kotlinx.coroutines.CoroutineDispatcher
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode.BACKGROUND
+import org.wordpress.android.models.ReaderPost
+import org.wordpress.android.modules.IO_THREAD
+import org.wordpress.android.ui.reader.actions.ReaderPostActions
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryEvent
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryEvent.PostLikeEnded
+import org.wordpress.android.util.EventBusWrapper
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class PostLikeActionUseCase @Inject constructor(
+    @Named(IO_THREAD) private val ioDispatcher: CoroutineDispatcher,
+    private val eventBusWrapper: EventBusWrapper
+) : ReaderRepositoryDispatchingUseCase(ioDispatcher) {
+    private val continuations: MutableMap<PostLikeRequest, Continuation<ReaderRepositoryEvent>?> = mutableMapOf()
+
+    init {
+        eventBusWrapper.register(this)
+    }
+
+    suspend fun perform(
+        post: ReaderPost,
+        isAskingToLike: Boolean,
+        wpComUserId: Long
+    ): ReaderRepositoryEvent {
+        val request = PostLikeRequest(post.postId, post.blogId, isAskingToLike, wpComUserId)
+
+        if (continuations[request] != null) {
+            throw IllegalStateException("Perform Like action has already been sent for this request")
+        }
+
+        return suspendCoroutine { cont ->
```

Comment:
I'm wondering if we need to use `suspendCancellableCoroutine` here. We are calling `super.stop()`, which invokes `parentJob.cancel()`. But I'm not sure it'll have any effect when the suspendCoroutine isn't cancelable. Tbh I'm not too familiar with these concepts. Let me know what you think.

Reply:
Yeah, using a continuation is new to me as well. I believe using the `suspendCancellableCoroutine` gives us a chance to do something like `continuation.invokeOnCancellation { println("ended because of cancel") }`, but not sure if we need that makes a difference here. I am going to change it to `suspendCancellableCoroutine` and do some more research.

Reply:
Okay, switched it out in f0a984a